monai/data/dataset.py

```diff
@@ -93,22 +97,67 @@ def __init__(self, data, transform, cache_num=sys.maxsize, cache_rate=1.0):
                 will take the minimum of (cache_num, data_length x cache_rate, data_length).
             cache_rate (float): percentage of cached data in total, default is 1.0 (cache all).
                 will take the minimum of (cache_num, data_length x cache_rate, data_length).
+            num_workers (int): the number of worker processes to use.
+                If 0 a single thread will be used. Default is 0.
         """
         if not isinstance(transform, Compose):
             transform = Compose(transform)
         super().__init__(data, transform)
         self.cache_num = min(cache_num, int(len(self) * cache_rate), len(self))
         self._cache = list()
         print('Load and cache transformed data...')
-        for i in range(self.cache_num):
-            process_bar(i + 1, self.cache_num)
-            item = data[i]
-            for _transform in transform.transforms:
-                # execute all the deterministic transforms before the first random transform
-                if isinstance(_transform, Randomizable):
-                    break
-                item = apply_transform(_transform, item)
-            self._cache.append(item)
+        if num_workers > 0:
+            self._item_processed = 0
+            lock = Lock()
+            if mode == "thread":
+                def _update_bar(arg):
+                    self._item_processed += 1
+                    lock.acquire()
+                    try:
+                        process_bar(self._item_processed, self.cache_num)
+                    finally:
+                        lock.release()
+                with ThreadPoolExecutor(max_workers=num_workers) as executor:
+                    for i in range(self.cache_num):
+                        future = executor.submit(self._load_cache_item_thread, i, transform, data)
+                        future.add_done_callback(_update_bar)
+            elif mode == "process":
+                def _add_item(item):
+                    self._cache.append(item)
+                    self._item_processed += 1
+                    lock.acquire()
```

Comment:
I think it's not correct to use a `thread lock` in this `multi-process` logic.
Please feel free to correct me if I misunderstand anything.
Thanks.

Reply:
Oh yes of course, it's a separate memory space. I'm fixing it now

Reply:
I ran again the timing tests just in case and multi-threads are still faster.

Reply:
Thanks for your update.
I tried to run your program in the [spleen segmentation notebook](https://github.com/gml16/MONAI/blob/309-cachedataset-multithread/examples/notebooks/spleen_segmentation_3d.ipynb)
The multi-thread version takes 95.75s to cache data, but the multi-process version still can't show the process bar. Could you please help double-check the process bar?
Thanks.